File: src/main/arraylist/cpp/exec_time.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <string>
#include <chrono>
#include <stdexcept>
#include "ArrayList.cpp"
#include "ArrayList.h"

using namespace std;
using namespace chrono;

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Uso: " << argv[0] << " <arquivo_de_entrada>" << endl;
        return 1;
    }

    ifstream entrada(argv[1]);
    if (!entrada.is_open()) {
        cerr << "Erro ao abrir o arquivo de entrada." << endl;
        return 1;
    }

    ArrayList arraylist;
    string linha;
    int valor;

    while (getline(entrada, linha)) {
        stringstream ss(linha);
        string operacao;
        ss >> operacao;

        if (operacao == "add_first") {
            ss >> valor;
            auto inicio = high_resolution_clock::now();
            arraylist.addFirst(valor);
            auto fim = high_resolution_clock::now();
            cout << "add_first: " << duration<double, micro>(fim - inicio).count() << " us" << endl;
        }
        else if (operacao == "add_last") {
            ss >> valor;
            auto inicio = high_resolution_clock::now();
            arraylist.addLast(valor);
            auto fim = high_resolution_clock::now();
cout << "add_last: " << duration<double, micro>(fim - inicio).count() << " us" << endl;
        }
        else if (operacao == "add_middle") {
            ss >> valor;
            int index = arraylist.getSize() / 2;
            auto inicio = high_resolution_clock::now();
            try {
                arraylist.addIndex(valor, index);
            } catch (const out_of_range& e) {}
            auto fim = high_resolution_clock::now();
            cout << "add_middle: " << duration<double, micro>(fim - inicio).count() << " us" << endl;
        }
        else if (operacao == "remove_first") {
            auto inicio = high_resolution_clock::now();
            try {
                arraylist.rmvFirst();
            } catch (const out_of_range& e) {}
            auto fim = high_resolution_clock::now();
            cout << "remove_first: " << duration<double, micro>(fim - inicio).count() << " us" << endl;
        }
        else if (operacao == "remove_last") {
            auto inicio = high_resolution_clock::now();
            try {
                arraylist.rmvLast();
            } catch (const out_of_range& e) {}
            auto fim = high_resolution_clock::now();
            cout << "remove_last: " << duration<double, micro>(fim - inicio).count() << " us" << endl;
        }
        else if (operacao == "remove_middle") {
            int index = arraylist.getSize() / 2;
            auto inicio = high_resolution_clock::now();
            try {
                arraylist.rmvIndex(index);
            } catch (const out_of_range& e) {}
            auto fim = high_resolution_clock::now();
            cout << "remove_middle: " << duration<double, micro>(fim - inicio).count() << " us" << endl;
        }
        else if (operacao == "get_first") {
            auto inicio = high_resolution_clock::now();
            try {
                arraylist.getFirst();
            } catch (const out_of_range& e) {}
            auto fim = high_resolution_clock::now();
            cout << "get_first: " << duration<double, micro>(fim - inicio).count() << " us" << endl;
        }
        else if (operacao == "get_last") {
            auto inicio = high_resolution_clock::now();
            try {
                arraylist.getLast();
            } catch (const out_of_range& e) {}
            auto fim = high_resolution_clock::now();
            cout << "get_last: " << duration<double, micro>(fim - inicio).count() << " us" << endl;
        }
        else if (operacao == "get_middle") {
            int index = arraylist.getSize() / 2;
            auto inicio = high_resolution_clock::now();
            try {
                arraylist.get(index);
            } catch (const out_of_range& e) {}
            auto fim = high_resolution_clock::now();
            cout << "get_middle: " << duration<double, micro>(fim - inicio).count() << " us" << endl;
        }
    }

    entrada.close();
    return 0;
}

File: src/main/arraylist/python/ExecTime.py
import time
import sys
import os
from statistics import median

from ArrayList import ArrayList 

class ExecTime:
    """
    Classe responsável por medir e registrar os tempos de execução de operações em uma ArrayList.
    Os resultados são salvos em arquivos separados para cada tipo de operação.
    """

    METHOD_NAMES = [
        "rmv_last",
        "add_last",
        "get_last",
        "rmv_first",
        "add_first",
        "get_first",
        "rmv_middle",
        "add_middle",
        "get_middle"
    ]

    @staticmethod
    def main():
        if len(sys.argv) < 2:
            print("ERRO: número de argumentos inválido!")
            return

        input_file = sys.argv[1]
        output_dir = "data/results/time/"

        try:
            os.makedirs(output_dir, exist_ok=True)

            writers = []
            try:
                for method in ExecTime.METHOD_NAMES:
                    file_path = f"{output_dir}{method}.data"
                    f = open(file_path, 'a+')
                    if os.fstat(f.fileno()).st_size == 0:
                        f.write("estrutura_linguagem tempo tamanho\n")
                    writers.append(f)
            except IOError as e:
                for f in writers:
                    f.close()
                raise

            with open(input_file, 'r') as reader:
                total_lines = sum(1 for line in reader if line.strip())

            with open(input_file, 'r') as reader:
                current_line = 0
                for line in reader:
                    line = line.strip()
                    if not line:
                        continue

                    current_line += 1
                    print(f"Processando linha {current_line}/{total_lines}", end='\r')

                    elements = list(map(int, line.split()))
                    ExecTime.test_arraylist_operations(writers, elements)

            print(f"\rConcluído! Processadas {total_lines} linhas.")

        except IOError as e:
            print(f"Erro ao processar arquivos: {e}")
        finally:
            for writer in writers:
                writer.close()

    @staticmethod
    def test_arraylist_operations(writers, elements):
        length = len(elements)
        middle = length // 2
        RUNS = 30
        all_times = [[] for _ in range(len(ExecTime.METHOD_NAMES))]

        # Como a ArrayList é imutável, precisamos reatribuir em cada operação
        array_list = ArrayList(elements)

        for run in range(RUNS):
            # Testa rmvLast
            start_time = time.perf_counter_ns()
            new_list = array_list.rmvLast()
            end_time = time.perf_counter_ns()
            all_times[0].append(end_time - start_time)
            array_list = ArrayList(new_list)  # Atualiza com o novo estado

            # Testa addLast
            start_time = time.perf_counter_ns()
            new_list = array_list.addLast(999)
            end_time = time.perf_counter_ns()
            all_times[1].append(end_time - start_time)
            array_list = ArrayList(new_list)

            # Testa getLast
            start_time = time.perf_counter_ns()
            array_list.getLast()
            end_time = time.perf_counter_ns()
            all_times[2].append(end_time - start_time)

            # Testa rmvFirst
            start_time = time.perf_counter_ns()
            new_list = array_list.rmvFirst()
            end_time = time.perf_counter_ns()
            all_times[3].append(end_time - start_time)
            array_list = ArrayList(new_list)

            # Testa addFirst
            start_time = time.perf_counter_ns()
            new_list = array_list.addFirst(999)
            end_time = time.perf_counter_ns()
            all_times[4].append(end_time - start_time)
            array_list = ArrayList(new_list)

            # Testa getFirst
            start_time = time.perf_counter_ns()
            array_list.getFirst()
            end_time = time.perf_counter_ns()
            all_times[5].append(end_time - start_time)

            # Testa rmvIndex (middle)
            start_time = time.perf_counter_ns()
            new_list = array_list.rmvIndex(middle)
            end_time = time.perf_counter_ns()
            all_times[6].append(end_time - start_time)
            array_list = ArrayList(new_list)

            # Testa addIndex (middle)
            start_time = time.perf_counter_ns()
            new_list = array_list.addIndex(999, middle)
            end_time = time.perf_counter_ns()
            all_times[7].append(end_time - start_time)
            array_list = ArrayList(new_list)

            # Testa get (middle)
            start_time = time.perf_counter_ns()
            array_list.get(middle)
            end_time = time.perf_counter_ns()
            all_times[8].append(end_time - start_time)

        for i, method in enumerate(ExecTime.METHOD_NAMES):
            if all_times[i]:
                median_time = int(median(all_times[i]))
                writers[i].write(f"arraylist_python {median_time} {length}\n")
                writers[i].flush()

if __name__ == "__main__":
    ExecTime.main()
